### ISSUE TITLE
world-based API - setup() -> World, update(world), draw(world)

### ISSUE THREAD

**Issue by acbart** [lordmauve/pgzero#174]
Hello! First, I want to compliment your library. It seems like you have a really cool tool here, and it's great to see how it's being used in real classrooms. I only found this recently, so I am still learning about the system.

I believe I am outside of your usual context, but I teach introductory Computer Science at the undergrad level. Although Pygame could be suitable for my students, I think a streamlined library like Pygame Zero has a big place in allowing simple game development earlier in the semester.

However, a major concern I have for this architecture from a pedagogical perspective is it's reliance on the global namespace. I agree this is fine in more basic contexts (e.g. K-12), it is problematic in more advanced settings. Simply put, I think it's bad to teach students to use global variables. I know pedagogy is not necessarily on the projects values, but I'd like to at least raise this discussion.

Global variables are generally not a good practice. They introduce many potential errors; if students neglect the globals keyword, there's a lot they could mess up with assignment. I know many colleagues at this level who would never encourage their use.

An alternative design (perhaps alongside the existing one or as a fork) would be based loosely off the architecture of the HTDP Universe library. There, they pass around World objects which can serve as a namespace. This encourages thinking concretely about the design of the game and enforcing a more concrete data flow.

Basically, there would be a new function and modifications to draw and update:

```python
setup() -> world
draw([world], [screen]) -> None
update([dt], [world]) -> None
```

Nothing actually needs to be done with these other than allow the world namespace to be passed around in a directed manner. This would encourage students to write code like this:

```python
def update(world):
  world.alien.left += 1
```

At this point, I've probably proposed too different an API and it doesn't seem to be in line with the design principles from your documentation. But the current architecture prohibits me from using Pygame Zero too seriously in my course, since it conflicts with all my lessons about avoiding global scope. So I hope you'll consider my thoughts.

Compliments again on your library!

**Comment by lordmauve** [comment]
In principle you're right, this kind of state management is better practice. The API is simple, and I like the idea of using Pygame Zero's "optional parameters" approach to hide this for beginners who will be happy with globals.

In practice, I believe we *have* to start with globals to make Pygame Zero accessible to the youngest kids. Also it aligns with the code they will be taught in plain Python, and on Micropython and so on. Then the problem is that we have already introduced the use of global state and there's really minimal incentive to switch and strictly adhere to the pattern you propose.

I also note that it's a hairs breadth away from a class-based API, which is an improvement because you can have many implementations - and thus naturally segues into the advantages of using classes for managing state.

```python
class Game:
    def __init__(self):
    def update(self, dt):
    def draw(self, screen):
    def on_key_down(self, ...):
```

I think I'd prefer to see the option to use a class-based API as the logical extension from the "global namespace" approach we promote for complete beginners.

**Comment by acbart** [comment]
Others might have different perspectives, but I will voice my own experience last semester - the transition to OO is a tough one. Particularly in Python, my students struggled with the concept of "self" and how to even organize their code. Although I think game design offers a great path to teach OO compared to many others contexts, I do think OO comes with a bit of a cognitive burden. If my goal was to teach OO, then having such an API is excellent. If my goal was to teach something else (e.g., functions, mutable state), then I would see it as a disadvantage to have to teach OO (because it'd add a host of new learning objectives with a tough overhead). There's some variability here at the undergrad level, about whether to teach Object-Oriented Programming.

I definitely agree that the global approach makes things more approachable to your primary audience, so I think that decision makes sense. I would still like to put in a plug for optional state parameters - four months later on the other side of teaching a full CS1 in Python, that still seems like the design that makes the most sense for my context. But again, I recognize that this project isn't necessarily meant for folks in my context :)
